status-network-contracts/docs/rln.md

```diff
@@ -0,0 +1,157 @@
+# RLN (Rate-Limiting Nullifier)
+
+## Overview
+
+The RLN (Rate-Limiting Nullifier) contract is a privacy-preserving identity registry that manages a set of identity
+commitments for the protocol participants. It enables users to register their identities, and allows authorized accounts
+to slash misbehaving users through a commit-reveal scheme that prevents front-running attacks.
+
+## Features
+
+- **Identity Registry:**
+
+  - Maintains a fixed-size registry of identity commitments determined by merkle tree depth.
+  - Registrations are sequential and tracked with unique indices.
+  - Prevents duplicate registrations of the same identity commitment.
+  - Supports privacy by storing only cryptographic commitments, not actual identities.
+
+- **Cryptographic Foundation:**
+
+  - Uses the Poseidon hash function to derive identity commitments from private keys.
+
+- **Commit-Reveal Slashing:**
+
+  - Two-step slashing mechanism prevents front-running attacks.
+  - Slashers commit to a slash operation with a cryptographic hash.
+  - Reveals must occur after a configurable reveal window, enforcing time-lock constraints.
+  - Subsequent slashes on the same account are queued and scheduled for future reveal windows.
+
+- **Reveal Window Management:**
+
+  - Configurable reveal window duration (default 1 hour, range 1 second to 1 day).
+  - Multiple commits for the same account are automatically queued.
+  - System tracks when the last reveal window started for each account.
+
+- **Karma Token Integration:**
+
+  - Integrates with the Karma contract to enforce economic penalties for violations.
+  - Slashing burns a percentage of the user's total Karma balance (both actual and virtual tokens).
+  - Slashing rewards are minted to a recipient specified as parameter.
+
+- **Role-Based Access Control:**
+
+  - Supports multiple roles: DEFAULT_ADMIN_ROLE, REGISTER_ROLE, and SLASHER_ROLE.
+  - Ensures only authorized accounts can register identities and execute slashes.
+
+- **Upgradeable Design:**
+  - Uses UUPS (Universal Upgradeable Proxy Standard) for secure contract upgrades.
+  - Initialized with owner, slasher, register, merkle tree depth, Karma token addresses, Poseidon hasher contract
+    address.
+
+## Identity Registration
+
+The RLN contract maintains a registry of identity commitments that represent encrypted representations of user
+identities.
+
+### Registration Process
+
+When an authorized account with REGISTER_ROLE registers an identity commitment:
+
+1. **Commitment Validation**: The contract verifies that the identity commitment hasn't been previously registered.
+2. **Registry Space**: The contract checks that the registry has capacity (not full based on merkle tree depth).
+3. **Commitment Storage**: The identity commitment is stored with the registrant's address and an incrementing index.
+4. **Event Emission**: A `MemberRegistered` event is emitted containing the commitment and its registry index.
```

Comment:
Also here, I don't think we need to elaborate on every single line of code. The event emission is irrelevant for someone reading these docs.